File: backend/app/schemas/spell_effect.py
"""
Spell Effect Pipeline schemas.

Defines the structured effect format for spell data.
Spells with an `effects` field use these types to describe
executable effects that SpellResolver can process.
"""
from typing import Optional, List, Dict, Any, Union, Literal
from pydantic import BaseModel, Field


# ── Trigger types ──────────────────────────────────────────────
TriggerType = Literal[
    "on_cast",               # 施法瞬间
    "on_hit",                # 攻击命中时（smite类）
    "on_weapon_hit",         # 武器命中时
    "on_reaction",           # 反应触发
    "on_target_downed",      # 目标生命降至 0
    "start_of_target_turn",  # 目标回合开始
    "end_of_target_turn",    # 目标回合结束
    "start_of_turn",         # 任意绑定主体回合开始
    "end_of_turn",           # 任意绑定主体回合结束
    "on_enter_zone",         # 进入区域
    "on_leave_zone",         # 离开区域
    "on_take_damage",        # 受伤时
    "on_concentration_end",  # 专注结束
    "on_action_invoked",     # 执行授予动作时
    "narrative",             # 纯RP/叙事型，无机械效果
]


# ── Target config ──────────────────────────────────────────────
class TargetConfig(BaseModel):
    type: Literal["self", "single", "multiple", "zone", "all_in_area"] = "single"
    count: Optional[Union[int, str]] = None  # 数字或 "spell_mod"
    filter: Optional[Dict[str, Any]] = None  # creature type / alignment / maxHP 等


# ── Save config ────────────────────────────────────────────────
class SaveConfig(BaseModel):
    ability: str  # "dex", "wis", "con", etc.
    on_success: Literal["no_effect", "half_damage", "partial"] = "no_effect"
    on_failure: Optional[str] = None  # 额外效果说明


# ── Escape config ──────────────────────────────────────────────
class EscapeConfig(BaseModel):
    trigger: str  # "end_of_turn", "on_take_damage", "action"
    method: Literal["save", "check", "auto_end"] = "save"
    ability: Optional[str] = None  # 用于 save/check 的属性


# ── Attack config ──────────────────────────────────────────────
class AttackConfig(BaseModel):
    type: Literal["melee_spell", "ranged_spell"] = "ranged_spell"
    on_miss: Literal["no_effect", "half_damage"] = "no_effect"
    # 多发射物（魔能爆束 / 灼热射线道）：每个发射物各掷一次独立法术攻击骰
    projectiles: int = 1
    projectilesAtCharacterLevel: Optional[Dict[str, int]] = None


# ── Scaling config ─────────────────────────────────────────────
class ScalingConfig(BaseModel):
    per_slot_above: int = 1           # 基础环位
    extra_dice: Optional[str] = None  # 每升一环加的骰子, e.g. "1d6"
    extra_targets: Optional[int] = None
    extra_duration: Optional[str] = None
    extra_value: Optional[int] = None  # 每升一环固定值, e.g. tempHp +5
    extra_projectiles: Optional[int] = None  # 每高于 per_slot_above 一环额外发射物数


# ── Effect primitives ─────────────────────────────────────────
class DealDamageEffect(BaseModel):
    type: Literal["deal_damage"] = "deal_damage"
    formula: str           # "8d6", "2d6+MOD"
    damage_type: str       # "fire", "radiant", etc.
    scaling: Optional[ScalingConfig] = None


class HealEffect(BaseModel):
    type: Literal["heal"] = "heal"
    formula: str           # "1d8+MOD"
    scaling: Optional[ScalingConfig] = None


class ApplyConditionEffect(BaseModel):
    type: Literal["apply_condition"] = "apply_condition"
    condition: str         # "paralyzed", "frightened", etc.
    condition_cn: Optional[str] = None
    escape: Optional[EscapeConfig] = None


class ModifyStatEffect(BaseModel):
    """Static stat modifier, e.g. AC +5, speed +10"""
    type: Literal["modify_stat"] = "modify_stat"
    stat: str              # "ac", "speed", "hp_max"
    formula: str           # "5", "13+DEX_MOD", "10"
    operation: Literal["add", "set"] = "add"


class ModifyRollEffect(BaseModel):
    """Roll modifier like Bless +1d4 to attack/save"""
    type: Literal["modify_roll"] = "modify_roll"
    roll_types: List[str]  # ["attack", "save"], ["ability_check"]
    formula: str           # "1d4", "PROF"
    operation: Literal["add", "subtract"] = "add"
    consume_on_use: bool = False  # 一次性修正（如导引之光的下次攻击优势）


class GrantTempHpEffect(BaseModel):
    type: Literal["grant_temp_hp"] = "grant_temp_hp"
    formula: str           # "5", "1d4+4"
    scaling: Optional[ScalingConfig] = None


class GrantResistanceEffect(BaseModel):
    type: Literal["grant_resistance"] = "grant_resistance"
    damage_types: List[str]  # ["bludgeoning", "piercing", "slashing"]


class GrantAdvantageEffect(BaseModel):
    type: Literal["grant_advantage"] = "grant_advantage"
    on: str                # "attack", "save", "ability_check", "incoming_attack"
    condition: Optional[Dict[str, Any]] = None  # {"ability": "dex"} etc.
    consume_on_use: bool = False  # 一次性效果


class ApplyEffectEffect(BaseModel):
    """Apply a full active_effect object (for complex effects)."""
    type: Literal["apply_effect"] = "apply_effect"
    effect_data: Dict[str, Any]  # 直接就是 active_effects 的一条


class NarrativeEffect(BaseModel):
    """Non-mechanical effect handled by DM/AI narration."""
    type: Literal["narrative"] = "narrative"
    description: Optional[str] = None


# Union of all effect primitives
EffectPrimitive = Union[
    DealDamageEffect,
    HealEffect,
    ApplyConditionEffect,
    ModifyStatEffect,
    ModifyRollEffect,
    GrantTempHpEffect,
    GrantResistanceEffect,
    GrantAdvantageEffect,
    ApplyEffectEffect,
    NarrativeEffect,
]


# ── Effect Phase ───────────────────────────────────────────────
class EffectPhase(BaseModel):
    """A single phase of spell execution."""
    trigger: TriggerType = "on_cast"
    target: Optional[TargetConfig] = None
    attack: Optional[AttackConfig] = None  # 法术攻击骰（近战/远程法术攻击）
    save: Optional[SaveConfig] = None
    effects: List[Dict[str, Any]] = Field(default_factory=list)  # 用 Dict 以保持 JSON 灵活性
    escape: Optional[EscapeConfig] = None
    duration: Optional[Dict[str, Any]] = None  # {"rounds": 10, "concentration": true}
    scaling: Optional[ScalingConfig] = None
    condition: Optional[Dict[str, Any]] = None  # 触发前置条件


# ── Resolve result ─────────────────────────────────────────────
class EffectResult(BaseModel):
    """Result of a single effect execution."""
    type: str
    target_token_id: Optional[int] = None
    target_name: Optional[str] = None
    # Damage/healing
    damage_dealt: int = 0
    healing_done: int = 0
    temp_hp_granted: int = 0
    # Condition
    condition_applied: Optional[str] = None
    condition_immune: bool = False
    # Save
    save_rolled: bool = False
    save_succeeded: Optional[bool] = None
    save_total: Optional[int] = None
    save_dc: Optional[int] = None
    # Attack
    attack_rolled: bool = False
    attack_hit: Optional[bool] = None
    attack_roll: Optional[int] = None   # d20 result (natural)
    attack_total: Optional[int] = None  # d20 + bonus
    target_ac: Optional[int] = None
    critical_hit: bool = False
    # Formula details
    formula_breakdown: Optional[str] = None
    # Misc
    description: Optional[str] = None


class SpellResolveResult(BaseModel):
    """Full result of spell resolution through the pipeline."""
    resolved: bool = True
    phase_results: List[List[EffectResult]] = Field(default_factory=list)
    # Aggregated
    total_damage: int = 0
    total_healing: int = 0
    narrative_parts: List[str] = Field(default_factory=list)
    # Effects applied (for WebSocket broadcast)
    effects_applied: List[Dict[str, Any]] = Field(default_factory=list)
    concentration_set: bool = False
    # Items generated by spell effects (e.g. generate_item)
    items_generated: List[Dict[str, Any]] = Field(default_factory=list)
    # Caster tokens whose concentration was broken by dispel_magic
    concentration_broken_token_ids: List[int] = Field(default_factory=list)
    # Caster tokens whose concentration targets changed because a damaged target escaped/ended the effect
    concentration_touched_token_ids: List[int] = Field(default_factory=list)
    # Tokens whose runtime spell projections should be refreshed
    runtime_touched_token_ids: List[int] = Field(default_factory=list)

File: frontend/app/types/spellEffect.ts
/**
 * Spell Effect Pipeline types.
 *
 * Defines the structured effect format matching backend schemas.
 * Spells with an `effects` field use these types for executable effects.
 */

// ── Trigger types ─────────────────────────────────────────────
export type TriggerType =
  | 'on_cast'               // 施法瞬间
  | 'on_hit'                // 攻击命中时（smite类）
  | 'on_weapon_hit'         // 武器攻击命中时
  | 'on_reaction'           // 反应触发
  | 'on_target_downed'      // 被标记目标倒地时
  | 'start_of_turn'         // 回合开始
  | 'end_of_turn'           // 回合结束
  | 'start_of_target_turn'  // 目标回合开始
  | 'end_of_target_turn'    // 目标回合结束
  | 'on_enter_zone'         // 进入区域
  | 'on_leave_zone'         // 离开区域
  | 'on_take_damage'        // 受伤时
  | 'on_concentration_end'  // 专注结束
  | 'on_action_invoked'     // 运行时动作被调用
  | 'narrative'             // 纯RP/叙事型，无机械效果

// ── Target config ─────────────────────────────────────────────
export interface TargetConfig {
  type: 'self' | 'single' | 'multiple' | 'zone' | 'all_in_area'
  count?: number | string   // number or "spell_mod"
  filter?: Record<string, unknown>
}

// ── Save config ───────────────────────────────────────────────
export interface SaveConfig {
  ability: string           // "dex", "wis", "con", etc.
  onSuccess: 'no_effect' | 'half_damage' | 'partial'
  onFailure?: string
}

// ── Attack config ─────────────────────────────────────────────
export interface AttackConfig {
  type: 'melee_spell' | 'ranged_spell'
  onMiss: 'no_effect' | 'half_damage'
}

// ── Escape config ─────────────────────────────────────────────
export interface EscapeConfig {
  trigger: string           // "end_of_turn", "on_take_damage", "action"
  method: 'save' | 'check' | 'auto_end'
  ability?: string
}

// ── Scaling config ────────────────────────────────────────────
export interface ScalingConfig {
  perSlotAbove: number      // 基础环位
  extraDice?: string        // 每升一环加的骰子, e.g. "1d6"
  extraTargets?: number
  extraDuration?: string
  extraValue?: number       // 每升一环固定值, e.g. tempHp +5
}

// ── Effect primitives ─────────────────────────────────────────
export interface DealDamageEffect {
  type: 'deal_damage'
  formula: string           // "8d6", "2d6+MOD"
  damageType: string        // "fire", "radiant"
  scaling?: ScalingConfig
}

export interface HealEffect {
  type: 'heal'
  formula: string
  scaling?: ScalingConfig
}

export interface ApplyConditionEffect {
  type: 'apply_condition'
  condition: string         // "paralyzed", "frightened"
  conditionCn?: string
  escape?: EscapeConfig
}

export interface ModifyStatEffect {
  type: 'modify_stat'
  stat: string              // "ac", "speed", "hp_max"
  formula: string           // "5", "13+DEX_MOD"
  operation: 'add' | 'set'
}

export interface ModifyRollEffect {
  type: 'modify_roll'
  rollTypes: string[]       // ["attack", "save"]
  formula: string           // "1d4"
  operation: 'add' | 'subtract'
  consumeOnUse?: boolean    // 一次性修正
}

export interface GrantTempHpEffect {
  type: 'grant_temp_hp'
  formula: string
  scaling?: ScalingConfig
}

export interface GrantResistanceEffect {
  type: 'grant_resistance'
  damageTypes: string[]
}

export interface GrantAdvantageEffect {
  type: 'grant_advantage'
  on: string                // "attack", "save", "incoming_attack"
  condition?: Record<string, unknown>
  consumeOnUse?: boolean    // 一次性效果
}

export interface GrantDisadvantageEffect {
  type: 'grant_disadvantage'
  on: string
  condition?: Record<string, unknown>
  consumeOnUse?: boolean
}

export interface GrantImmunityEffect {
  type: 'grant_immunity'
  damageTypes?: string[]
  conditions?: string[]           // 免疫的状态，如 ["frightened","paralyzed"]
}

export interface ApplyMarkEffect {
  type: 'apply_mark'
  params?: Record<string, unknown>
}

export interface RetargetMarkEffect {
  type: 'retarget_mark'
}

export interface ConditionalExtraDamageEffect {
  type: 'conditional_extra_damage'
  formula: string
  damageType: string
}

export interface SetRuntimeParamEffect {
  type: 'set_runtime_param'
  key: string
  value: unknown
}

export interface ClearRuntimeParamEffect {
  type: 'clear_runtime_param'
  key: string
}

export interface EndSpellInstanceEffect {
  type: 'end_spell_instance'
}

export interface ApplyEffectEffect {
  type: 'apply_effect'
  effectData: Record<string, unknown>
}

export interface NarrativeEffect {
  type: 'narrative'
  description?: string
}

// ── UI / Visual Effect Primitives (功能中间件 — UI 效果) ─────────
export interface ResizeTokenEffect {
  type: 'resize_token'
  sizeDelta: number              // +1 变大, -1 缩小
}

export interface ApplyTokenFilterEffect {
  type: 'apply_token_filter'
  filter: {
    blur?: number
    opacity?: number
    glow?: string
    glowRadius?: number
    glowAnimation?: 'pulse'
    overlay?: string
    saturate?: number
    brightness?: number
  }
}

export interface SetVisibilityEffect {
  type: 'set_visibility'
  mode: 'invisible' | 'ethereal'
  filter?: ApplyTokenFilterEffect['filter']
}

export interface SetDisguiseEffect {
  type: 'set_disguise'
  disguiseType: 'appearance' | 'form_change'
  requiresImage: boolean
}

export interface SpawnIllusionEffect {
  type: 'spawn_illusion'
  illusionType: string           // 'visual' | 'auditory' | 'visual_auditory'
  controllable?: boolean
  physicalPass?: boolean
}

export interface CreateZoneVisualEffect {
  type: 'create_zone_visual'
  zoneType: string               // 'fog' | 'darkness' | 'light' | 'terrain' | 'other'
  obscurement?: 'heavy' | 'light'
  difficultTerrain?: boolean
  spreadsAroundCorners?: boolean
  windDispel?: Record<string, number>  // {"moderate": 1, "strong": 3}
  cloudMovement?: { distance: number; direction: string; timing: string }
}

export interface ApplyTransformationEffect {
  type: 'apply_transformation'
  transformType: string          // 'polymorph' | 'wild_shape' | 'alter_self' | 'animal_shapes'
}

// ── Illumination Primitive (光照/黑暗效果) ──────────────────────
export interface ApplyIlluminationEffect {
  type: 'apply_illumination'
  lightType: 'light' | 'darkness'
  brightRadius: number             // 明亮光照半径（尺）
  dimRadius: number                // 微光半径（尺）
  darknessRadius?: number          // 黑暗半径（仅 darkness 类型）
  attachTo: 'object' | 'caster' | 'point' | 'target'
  color?: string                   // 光照颜色
  blocksDarkvision?: boolean       // 是否阻挡暗视觉
  dispelsLevel?: number            // 能驱散的对立法术最高环位
  isSunlight?: boolean             // 是否算作日光
  movable?: boolean
  moveAction?: string              // 'action' | 'bonus_action' | 'free'
}

// ── Summoning Primitive (召唤效果) ──────────────────────────────
export interface SpawnSummonEffect {
  type: 'spawn_summon'
  monsterId?: string               // 预设怪物 ID（如 riding_horse）
  instanceName: string             // 显示名称
  tokenSize?: string               // "1x1" | "2x2" 等
  avatarPath?: string              // 头像路径
  hpFormula?: string               // 生命值公式（如 "5d8"）
  count?: number                   // 召唤数量
  faction?: 'player' | 'neutral'   // 阵营
  description?: string
}

// ── Condition Removal Primitive (解除状态效果) ──────────────────
export interface RemoveConditionEffect {
  type: 'remove_condition'
  conditions: string[]             // 可解除的状态列表，如 ["blinded","deafened","poisoned"]
  mode: 'choose_one' | 'all'      // 选一个移除 | 全部移除
}

// ── Movement / Teleportation Primitives (位移/移动效果) ──────────
export interface TeleportEffect {
  type: 'teleport'
  range: number                  // 传送距离（尺）
  mode: 'self' | 'target' | 'swap'  // 自身传送 | 目标传送 | 互换位置
  mustSee?: boolean              // 是否需要视线
  carriesOthers?: boolean        // 是否可携带他人
}

export interface ModifyMovementEffect {
  type: 'modify_movement'
  movementType: string           // 'walk' | 'fly' | 'swim' | 'climb' | 'burrow'
  formula?: string               // 速度值或公式，如 "60" 或 "walk_speed"
  operation: 'set' | 'add' | 'grant'  // 设置 | 增减 | 授予新移动方式
  hover?: boolean                // 飞行是否悬浮
}

export interface RestrictMovementEffect {
  type: 'restrict_movement'
  restriction: string            // 'immobilized' | 'halved' | 'zero' | 'no_teleport'
  escape?: EscapeConfig
}

// ── Grant Action (法术授予的可执行动作) ─────────────────────────
export interface GrantActionEffect {
  type: 'grant_action'
  actionType: string              // "action", "bonus_action", "reaction"
  actionName: string              // 中文名，如 "灵体武器攻击"
  actionNameEn?: string           // 英文名
  icon?: string                   // emoji icon
  actionKind: string              // "melee_spell_attack", "dash", "command" 等
  attack?: AttackConfig
  save?: SaveConfig
  damage?: { formula: string; damageType: string }
  healing?: { formula: string }
  movement?: { range: number }
  areaOfEffect?: { type?: string; size?: number }
  attackCount?: number
  commandRange?: number
  allowedActions?: string[]
  triggerCondition?: string
  scaling?: ScalingConfig
  description?: string
}

// ── Forced Movement Primitive (强制位移) ────────────────────────
export interface ForcedMovementEffect {
  type: 'forced_movement'
  direction: 'push' | 'pull' | 'toward_point'  // 推开 | 拉近 | 朝向指定点
  distance: number                               // 距离（尺）
  relativeTo?: 'caster' | 'spell_origin'        // 相对于谁
}

// ── Wall Creation Primitive (墙体) ─────────────────────────────
export interface CreateWallEffect {
  type: 'create_wall'
  shape: 'line' | 'ring'
  lengthFt?: number                              // 线形总长
  heightFt?: number
  thicknessFt?: number
  hpPerSegment?: number                          // 每段 HP (null=不可摧毁)
  blocksMovement?: boolean
  blocksLineOfSight?: boolean
  blocksProjectiles?: boolean
  passThroughDamage?: { formula: string; damageType: string }
}

// ── Barrier Primitive (力场屏障) ───────────────────────────────
export interface CreateBarrierEffect {
  type: 'create_barrier'
  shape: 'sphere' | 'cube' | 'dome' | 'cage'
  radiusFt?: number
  sideFt?: number                                // cube 边长
  blocksMagic?: boolean
  blocksPhysical?: boolean
  indestructible?: boolean
  teleportBlocked?: boolean
}

// ── Moving Aura Primitive (跟随光环) ──────────────────────────
export interface CreateMovingAuraEffect {
  type: 'create_moving_aura'
  radiusFt: number
  anchor: 'caster' | 'target'
  auraType?: string                              // 'damage' | 'buff' | 'debuff' | 'protection'
}

// ── Counter Spell Primitive (法术反制) ────────────────────────
export interface CounterSpellEffect {
  type: 'counter_spell'
  autoCounterLevel: number                       // 自动反制的最高环位
  checkAbility?: string                          // 反制高环法术时的属性
  checkDcFormula?: string                        // DC 公式 "10 + spell_level"
}

// ── Grant Sense Primitive (感知能力) ─────────────────────────
export interface GrantSenseEffect {
  type: 'grant_sense'
  senseType: string                              // 'truesight' | 'blindsight' | 'tremorsense' | 'detect_invisible' | 'detect_magic' | 'remote_sensor'
  range?: number                                 // 感知范围（尺）
}

// ── Small utility primitives ─────────────────────────────────
export interface PreventHealingEffect {
  type: 'prevent_healing'
  duration?: string                              // 'until_caster_next_turn' | 'rounds'
}

export interface StabilizeEffect {
  type: 'stabilize'
}

export interface InstantKillEffect {
  type: 'instant_kill'
  hpThreshold: number                            // HP <= 此值即死
}

export interface ResurrectEffect {
  type: 'resurrect'
  hpRestored?: string                            // 公式，如 "1" 或 "full"
  debuffDuration?: string                        // 减值持续时间
  requiresBody?: boolean
}

// ── Stored Trigger Primitive (延迟触发) ──────────────────────
export interface StoredTriggerEffect {
  type: 'stored_trigger'
  triggerCondition: string                       // 'enter_area' | 'take_damage' | 'custom'
  storedSpellId?: string
}

// ── Suppress Magic Primitive (压制魔法) ─────────────────────
export interface SuppressMagicEffect {
  type: 'suppress_magic'
  maxSpellLevel?: number                         // null = 全部
  scope: 'zone' | 'target'
}

export type EffectPrimitive =
  | DealDamageEffect
  | HealEffect
  | ApplyConditionEffect
  | ModifyStatEffect
  | ModifyRollEffect
  | GrantTempHpEffect
  | GrantResistanceEffect
  | GrantImmunityEffect
  | GrantAdvantageEffect
  | GrantDisadvantageEffect
  | ApplyMarkEffect
  | RetargetMarkEffect
  | ConditionalExtraDamageEffect
  | SetRuntimeParamEffect
  | ClearRuntimeParamEffect
  | EndSpellInstanceEffect
  | ApplyEffectEffect
  | NarrativeEffect
  | GrantActionEffect
  // UI / Visual effects
  | ResizeTokenEffect
  | ApplyTokenFilterEffect
  | SetVisibilityEffect
  | SetDisguiseEffect
  | SpawnIllusionEffect
  | CreateZoneVisualEffect
  | ApplyTransformationEffect
  // Condition removal
  | RemoveConditionEffect
  // Illumination / Summoning effects
  | ApplyIlluminationEffect
  | SpawnSummonEffect
  // Movement / Teleportation effects
  | TeleportEffect
  | ModifyMovementEffect
  | RestrictMovementEffect
  // Forced movement / Walls / Barriers
  | ForcedMovementEffect
  | CreateWallEffect
  | CreateBarrierEffect
  | CreateMovingAuraEffect
  // Counter / Sense / Utility
  | CounterSpellEffect
  | GrantSenseEffect
  | PreventHealingEffect
  | StabilizeEffect
  | InstantKillEffect
  | ResurrectEffect
  // Trigger / Suppress
  | StoredTriggerEffect
  | SuppressMagicEffect

// ── Effect Phase ──────────────────────────────────────────────
export interface EffectPhase {
  trigger: TriggerType
  target?: TargetConfig
  attack?: AttackConfig     // 法术攻击骰（近战/远程法术攻击）
  save?: SaveConfig
  effects: EffectPrimitive[]
  escape?: EscapeConfig
  duration?: { rounds?: number; minutes?: number; concentration?: boolean }
  scaling?: ScalingConfig
  condition?: Record<string, unknown>   // 触发前置条件
  // ── 生物类型过滤 ──
  targetCreatureTypes?: string[]        // 仅对这些类型生效 (如 ["humanoid"])
  excludeCreatureTypes?: string[]       // 对这些类型无效 (如 ["undead","construct"])
}
